storage/clickhousespanstore/writer.go

```diff
@@ -114,12 +133,30 @@ func (w *SpanWriter) backgroundWriter() {
 		}
 
 		if flush {
-			if err := w.writeBatch(batch); err != nil {
-				w.logger.Error("Could not write a batch of spans", "error", err)
+			if lastDone != nil {
+				if !*lastDone {
+					w.size = lastSize / 2
+					w.logger.Debug("Could now write spans on time, decreasing batch size", "size", w.size)
+				} else {
+					w.size = min(maxSize, max(lastSize+sizeIncrement, w.size))
+					if w.size > maxSize {
+						w.size = maxSize
+					}
+					w.logger.Debug("Increasing batch size", "size", w.size)
+					lastDone = nil
+				}
 			}
+			last = time.Now()
+			lastDone = new(bool)
+			lastSize = int64(len(batch))
+			go func() {
```

Comment:
why are you using goroutine here? Could you please explain why the flush/write is done after increasing the size? 

I would imagine that batch size should be changed after the flush/write.

Reply:
> why are you using goroutine here?

It will make timer flushes more predictable, 'cause write time won't affect it.

> I would imagine that batch size should be changed after the flush/write.

I thought to check whether the previous flush was done in time before current flush, but it was too complicated.

Reply:
with the flush in a separate goroutine this can happen:
* two concurrent calls `w.writeBatch` 
* concurrent modifications of the parameters (e.g. size)